src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/FindAllReferences/FindAllReferencesEndpoint.cs

```diff
@@ -0,0 +1,157 @@
+        {
+            const string AssociatedServerCapability = "referencesProvider";
+
+            var registrationOptions = new ReferenceOptions()
+            {
+                // TODO: GH issue link to support progress 
+                WorkDoneProgress = false,
+            };
+
+            return new RegistrationExtensionResult(AssociatedServerCapability, new SumType<bool, ReferenceOptions>(registrationOptions));
+        }
+
+        protected override string CustomMessageTarget => RazorLanguageServerCustomMessageTargets.RazorReferencesEndpointName;
+
+        protected override Task<IDelegatedParams?> CreateDelegatedParamsAsync(ReferenceParamsBridge request, RazorRequestContext requestContext, Projection projection, CancellationToken cancellationToken)
+        {
+            var documentContext = requestContext.GetRequiredDocumentContext();
+            return Task.FromResult<IDelegatedParams?>(new DelegatedPositionParams(
+                    documentContext.Identifier,
+                    projection.Position,
+                    projection.LanguageKind));
+        }
+
+        protected override async Task<VSInternalReferenceItem[]> HandleDelegatedResponseAsync(VSInternalReferenceItem[] delegatedResponse, ReferenceParamsBridge originalRequest, RazorRequestContext requestContext, Projection projection, CancellationToken cancellationToken)
+        {
+            var remappedLocations = new List<VSInternalReferenceItem>();
+
+            foreach (var referenceItem in delegatedResponse)
+            {
+                if (referenceItem?.Location is null || referenceItem.Text is null)
+                {
+                    continue;
+                }
+
+                // Temporary fix for codebehind leaking through
+                // Revert when https://github.com/dotnet/aspnetcore/issues/22512 is resolved
+                referenceItem.DefinitionText = FilterReferenceDisplayText(referenceItem.DefinitionText);
+                referenceItem.Text = FilterReferenceDisplayText(referenceItem.Text);
+
+                // Indicates the reference item is directly available in the code
+                referenceItem.Origin = VSInternalItemOrigin.Exact;
+
+                if (!_featureOptions.IsVirtualCSharpFile(referenceItem.Location.Uri) &&
+                    !_featureOptions.IsVirtualHtmlFile(referenceItem.Location.Uri))
+                {
+                    // This location doesn't point to a virtual file. No need to remap.
+                    remappedLocations.Add(referenceItem);
+                    continue;
+                }
+
+                var documentContext = requestContext.GetRequiredDocumentContext();
+                var codeDocument = await documentContext.GetCodeDocumentAsync(cancellationToken).ConfigureAwait(false);
+
+                if (!_documentMappingService.TryMapFromProjectedDocumentRange(codeDocument, referenceItem.Location.Range, out var documentRange))
```

Comment:
This looks wrong.. if you're getting the code document for the document context being operated on, then it will try to remap the range in the context of that document.. but the reference could come from a different razor document, right?

I think you want to call the overload that takes a uri and a range, and gives back both, like GTD does:
https://github.com/dotnet/razor/blob/34f1c25ea3cf93ca9e083f13353c028774cdfb99/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Definition/RazorDefinitionEndpoint.cs#L120

Reply:
Ooh good catch.i should also add a multifile test of possible

Reply:
Tested manually. Not sure what procedure for multifile test is but can add as a follow up to this

Reply:
Now that this is in, I'll unskip https://github.com/davidwengier/razor/blob/ComponentAttributes/src/Razor/test/Microsoft.VisualStudio.Razor.IntegrationTests/FindAllReferencesTests.cs#L52 in my branch which should cover it